### PR TITLE
Fixes #19826 - Use subset layer to manage fid in ogr provider when a filter is applied

### PR DESCRIPTION
**Pull request by pblottiere**
## Description

Fixes https://issues.qgis.org/issues/19826.

## Checklist

- [x] Commit messages are descriptive and explain the rationale for changes
- [x] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [ ] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [ ] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and contain sufficient information in the commit message to be documented
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [x] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTE.md#contributing-to-qgis) before each commit

### CODE REVIEW

File: tests/src/python/test_qgsserver_wms_getfeatureinfo.py [hunk 1]
```diff
@@ -259,7 +259,6 @@ def testGetFeatureInfoRelationReference(self):
                                  'test_project_values.qgz')
 
     # TODO make filter work with gpkg and move test inside testGetFeatureInfoFilter function
```
Comment (m-kuhn) [inline tests/src/python/test_qgsserver_wms_getfeatureinfo.py]: Is the TODO here still open or was that what you solved?

Comment (pblottiere) [reply tests/src/python/test_qgsserver_wms_getfeatureinfo.py]: Oh yes, you're right, I have to remove this comment... Thanks :)

**Comment by nyalldawson** [comment]
@pblottiere could you add a non server ogr provider test which specifically targets this issue?

**Comment by pblottiere** [comment]
@nyalldawson 

> could you add a non server ogr provider test which specifically targets this issue?

Done

**Comment by nyalldawson** [comment]
Just to understand - the bug happens when the source has first...fid true, yet the fid is a valid field in the data source?

**Comment by pblottiere** [comment]
> Just to understand - the bug happens when the source has first...fid true, yet the fid is a valid field in the data source?

Yes. And actually, these cases are already covered by default in the provider, but not when a subset string is applied (it's what we are fixing in this PR).

**Pull request merged**